packages/graph/hash_graph/lib/graph/src/store/postgres/query/database.rs

```diff
@@ -0,0 +1,272 @@
+use std::{borrow::Cow, fmt, fmt::Write};
+
+use serde::Serialize;
+
+use crate::store::postgres::query::Render;
+
+/// A table available in the database.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Table {
+    TypeIds,
+}
+
+impl Table {
+    /// Returns the [`Column`] used for joining another `Table` on this `Table`.
+    pub const fn source_join_column(self) -> Column<'static> {
+        let column = match self {
+            Self::TypeIds => "version_id",
+        };
+        Column {
+            table: self,
+            access: ColumnAccess::Table { column },
+        }
+    }
+
+    /// Returns the [`Column`] used for joining this `Table` on another `Table`.
+    pub const fn target_join_column(self) -> Column<'static> {
+        let column = match self {
+            Self::TypeIds => "version_id",
+        };
+        Column {
+            table: self,
+            access: ColumnAccess::Table { column },
+        }
+    }
+}
+
+impl Render for Table {
+    fn render(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_char('"')?;
+        self.serialize(&mut *fmt)?;
+        fmt.write_char('"')
+    }
+}
+
+/// A table available in a compiled query.
+///
+/// When joining tables, each table requires a unique name. A `TableRef` may either be a plain
+/// [`Table`], or a [`Table`] with additional information attached to uniquely identify the table.
+///
+/// # Examples
+///
+/// When specifying multiple conditions or deeply nested queries containing the same [`Table`],
+/// [`Alias`] uniquely identifies the condition and the depth of the query.
+///
+/// ## Multiple Conditions
+///
+/// When searching for a [`PropertyType`], which should contain two different [`DataType`]s
+/// specified by it's `$id`, `Table::DataTypes` has to be joined twice, but with different
+/// conditions.
```

Comment:
```suggestion
/// When searching for a [`PropertyType`] specified by its `$id`, if it contains references
/// to two different [`DataType`]s then `Table::DataTypes` has to be joined twice but with different
/// conditions.
```

Do you mean this? Or were you trying to say that you're joining the Data Types based on _their_ IDs?